index.js

```diff
@@ -44,90 +52,162 @@ function cmdShim (src, to, opts) {
 }
 
 function cmdShim_ (src, to, opts) {
-  return Promise.all([
-    rm(to),
-    rm(`${to}.ps1`),
-    opts.createCmdFile && rm(`${to}.cmd`)
-  ])
-    .then(() => writeShim(src, to, opts))
+  opts = Object.assign({}, DEFAULT_OPTIONS, opts)
+  return searchScriptRuntime(src).then((srcRuntimeInfo) => {
+    // Always tries to create all types of shims by calling `writeAllShims` as of now.
+    // Append your code here to change the behavior in response to `srcRuntimeInfo`.
+
+    // Create 3 shims for (Ba)sh in Cygwin / MSYS, no extension) & CMD (.cmd) & PowerShell (.ps1)
+    return writeAllShims(src, to, srcRuntimeInfo, opts)
+  })
 }
 
-function writeShim (src, to, opts) {
+function writeAllShims (src, to, srcRuntimeInfo, opts) {
   opts = Object.assign({}, DEFAULT_OPTIONS, opts)
-  const defaultArgs = opts.preserveSymlinks ? '--preserve-symlinks' : ''
-  // make a cmd file and a sh script
-  // First, check if the bin is a #! of some sort.
-  // If not, then assume it's something that'll be compiled, or some other
-  // sort of script, and just call it directly.
-  return mkdir(path.dirname(to))
-    .then(() => {
-      return fs.readFile(src, 'utf8')
-        .then(data => {
-          const firstLine = data.trim().split(/\r*\n/)[0]
-          const shebang = firstLine.match(shebangExpr)
-          if (!shebang) return writeShim_(src, to, Object.assign({}, opts, {args: defaultArgs}))
-          const prog = shebang[1]
-          const args = (shebang[2] && ((defaultArgs && (shebang[2] + ' ' + defaultArgs)) || shebang[2])) || defaultArgs
-          return writeShim_(src, to, Object.assign({}, opts, {prog, args}))
+  return Promise.all([
+    [generateShShim, '', true],
+    [generateCmdShim, '.cmd', opts.createCmdFile],
+    [generatePwshShim, '.ps1', opts.createPwshFile]
+  ].map(([generateShimScript, extension, shouldCreate]) => shouldCreate && writeShim(src, to + extension, srcRuntimeInfo, generateShimScript, opts)))
+}
+
+function writeShimPre (target) {
+  return rm(target).then(() => mkdir(path.dirname(target)))
+}
```

Comment:
The previous code was calling mkdir only once. This one will call mkdir for every shim. It means that this version performs more file system operations. Please call mkdir only once.

Reply:
All right, but I wonder where I should move it to...  I suppose in `cmdShim_` is the best.